components/webvr/webvr_thread.rs

```diff
@@ -0,0 +1,349 @@
+            Ok(display) => {
+                sender.send(Ok(display.borrow().inmediate_frame_data(near, far))).unwrap()
+            },
+            Err(msg) => sender.send(Err(msg.into())).unwrap()
+        }
+    }
+
+    fn handle_reset_pose(&mut self,
+                         pipeline: PipelineId,
+                         display_id: u64,
+                         sender: IpcSender<WebVRResult<VRDisplayData>>) {
+        match self.access_check(pipeline, display_id) {
+            Ok(display) => {
+                display.borrow_mut().reset_pose();
+                sender.send(Ok(display.borrow().data())).unwrap();
+            },
+            Err(msg) => {
+                sender.send(Err(msg.into())).unwrap()
+            }
+        }
+    }
+
+    fn access_check(&self, pipeline: PipelineId, display_id: u64) -> Result<&VRDisplayPtr, &'static str> {
+        if *self.presenting.get(&display_id).unwrap_or(&pipeline) != pipeline {
+            return Err("Device owned by another context");
+        }
+        self.service.get_display(display_id).ok_or("Device not found")
+    }
+
+    fn handle_request_present(&mut self,
+                              pipeline: PipelineId,
+                              display_id: u64,
+                              sender: IpcSender<WebVRResult<()>>) {
+        match self.access_check(pipeline, display_id).map(|d| d.clone()) {
+            Ok(display) => {
+                self.presenting.insert(display_id, pipeline);
+                let data = display.borrow().data();
+                sender.send(Ok(())).unwrap();
+                self.notify_event(VRDisplayEvent::PresentChange(data, true));
+            },
+            Err(msg) => {
+                sender.send(Err(msg.into())).unwrap();
+            }
+        }
+    }
+
+    fn handle_exit_present(&mut self,
+                         pipeline: PipelineId,
+                         display_id: u64,
+                         sender: Option<IpcSender<WebVRResult<()>>>) {
+        match self.access_check(pipeline, display_id).map(|d| d.clone()) {
+            Ok(display) => {
+                self.presenting.remove(&display_id);
+                if let Some(sender) = sender {
+                    sender.send(Ok(())).unwrap();
+                }
+                let data = display.borrow().data();
+                self.notify_event(VRDisplayEvent::PresentChange(data, false));
+            },
+            Err(msg) => {
+                if let Some(sender) = sender {
+                    sender.send(Err(msg.into())).unwrap();
+                }
+            }
+        }
+    }
+
+    fn handle_create_compositor(&mut self, display_id: u64) {
+        let compositor = self.service.get_display(display_id).map(|d| WebVRCompositor(d.as_ptr()));
+        self.vr_compositor_chan.send(compositor).unwrap();
+    }
+
+    fn poll_events(&mut self, sender: IpcSender<bool>) {
+        let events = self.service.poll_events();
+        if !events.is_empty() {
```

Comment:
Can more events come in while you are handling the current set? Elsewhere in Servo event polling threads, we use a pattern that is more like `while (events = poll_events() && !events.is_empty()) { ... }` to avoid falling into the polling delay interval during periods of rapid interactions / event floods.